scrapy_poet/page_input_providers.py

```diff
@@ -1,102 +1,127 @@
-        """You can override this method to receive external dependencies.
+    A Page Object Input Provider (POIP) takes responsibility for providing
+    instances of some types to Scrapy callbacks. The types a POIP provides must
+    be declared in the class attribute ``provided_classes``.
 
-        Currently, scrapy-poet is able to inject instances of the following
-        classes as *provider* dependencies:
+    POIPs are initialized when the spider starts by invoking the ``__init__`` method,
+    which receives the crawler instance as argument.
 
-        - :class:`~scrapy.http.Request`
-        - :class:`~scrapy.http.Response`
-        - :class:`~scrapy.crawler.Crawler`
-        - :class:`~scrapy.settings.Settings`
-        - :class:`~scrapy.statscollectors.StatsCollector`
+    The ``__call__`` method must be overridden, and it is inside this method
+    where the actual instances must be build. The default ``__call__`` signature
+    is as follows:
 
-        .. warning::
+    .. code-block:: python
 
-            Scrapy doesn't know when a Request is going to generate
-            a Response, a TextResponse, an HtmlResponse,
-            or any other type that inherits from Response.
-            Because of this,
-            you should always annotate your provider's response argument
-            with the Response type.
-            If your provider needs a TextResponse,
-            you need to validate it by yourself,
-            the same way you would need to do when using Scrapy callbacks.
-            Example:
+        def __call__(self, to_provide: Set[Callable]) -> Dict[Callable, Any]:
 
-            .. code-block:: python
+    Therefore, it receives a list of types to be provided and return a dictionary
+    with the instances created indexed by its type (don't get confused by the
+    ``Callable`` annotation. Think on it as a synonym of ``Type``).
 
-                @provides(MyCustomResponseData)
-                class MyCustomResponseDataProvider(PageObjectInputProvider):
+    Additional dependencies can be declared in the ``__call__`` signature
+    that will be automatically injected. Currently, scrapy-poet is able
+    to inject instances of the following classes:
 
-                    def __init__(self, response: Response):
-                        assert isinstance(response, TextResponse)
-                        self.response = response
-        """
+    - :class:`~scrapy.http.Request`
+    - :class:`~scrapy.http.Response`
+    - :class:`~scrapy.crawler.Crawler`
+    - :class:`~scrapy.settings.Settings`
+    - :class:`~scrapy.statscollectors.StatsCollector`
 
-    @abc.abstractmethod
-    def __call__(self):
-        """This method is responsible for building Page Input dependencies."""
+    Finally, ``__call__`` function can execute asynchronous code. Just
+    either prepend the declaration with ``async`` or annotate it with
+    ``@inlineCallbacks`` depending on how Scrapy ``TWISTED_REACTOR``
+    is configured.
```

Comment:
`async def` support shouldn't require setting asyncio reactor, see https://docs.scrapy.org/en/latest/topics/coroutines.html

Reply:
If we declare `async def` support for providers, it'd be nice to have a test for it. Maybe modify one of the inlineCallbacks + returnValue test, or copy it? Or maybe there is such test already, and I missed it :)

Reply:
TIL. I added the test case and clarified a little bit the documentation here: https://github.com/scrapinghub/scrapy-poet/pull/41/commits/adef92ed36287e57fb205dec8f140c8f0d0a0985